### PR TITLE
obs-ffmpeg: Fix m3u8 recording in AMF

### PR DESCRIPTION
**Pull request by shinji3**
### Description

SPS/PPS header insert

### Motivation and Context

When recording in m3u8 format using AMD HW H.264, the second and subsequent TS files are broken.
Fixed second and subsequent TS files are broken.

### How Has This Been Tested?

Tested with start recording in OBS Studio

### Types of changes

Bug fix (non-breaking change which fixes an issue)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by WizardCM** [comment]
Could you define "broken" for the subsequent files?

**Comment by shinji3** [comment]
Can't play TS files in VLC player

I uploaded a bug test video
https://github.com/shinji3/BugTestVideoAMF

File: plugins/obs-ffmpeg/texture-amf.cpp
Comment (derrod) [inline plugins/obs-ffmpeg/texture-amf.cpp:1161]: This should probably be only conditionally enabled like the implementation for NVENC/x264 is, see PR #3083 for reference.

**Review by derrod — DISMISSED** [review]

File: plugins/obs-ffmpeg/texture-amf.cpp
Comment (tytan652) [inline plugins/obs-ffmpeg/texture-amf.cpp:1014]: The property is not required, `obs_data` is enough. The PR should work without it.

**Comment by pkviet** [comment]
your code breaks recording into m3u8 for nvenc and x264: the first segment can't be played.
For file recording, header repetition for these encoders is not required. It breaks hls because it records in annex B bitstream, which the "repeat_headers" setting adds for nvenc and x264.
So two solutions:
- either use another name for the setting than 'repeat_headers', like 'amf_insert_headers'
- add a check against the id of the encoder ("h264_texture_amf")
For extra safety, you could do both.
I haven't been able to confirm the issue yet on an amd gpu. Will report later.

**Comment by pkviet** [comment]
Also, you didn't properly rebase to master. You shouldn't merge the new commits from master but do a proper rebase .
git fetch origin
git rebase -i origin/master 
So remove this ugly merge commit.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c
Comment (pkviet) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c:394]: change the name of this setting (ex: amf_repeat_headers) and put everything behind a check of the encoder id (h264_texture_amf or h265_texture_amf) (don't bother with the fallback encoders).

File: plugins/obs-ffmpeg/texture-amf.cpp
Comment (pkviet) [inline plugins/obs-ffmpeg/texture-amf.cpp:936]: change the setting name, say : amf_repeat_headers

File: plugins/obs-ffmpeg/texture-amf.cpp
Comment (pkviet) [inline plugins/obs-ffmpeg/texture-amf.cpp:1166]: change the setting name here too, say : amf_repeat_headers

**Review by pkviet — CHANGES_REQUESTED** [review]

**Comment by pkviet** [comment]
also, for future reference, @RytoEX unearthed this discussion :https://github.com/GPUOpen-LibrariesAndSDKs/AMF/issues/293
Without these infos, it would have been impossible to guess that HEADER_INSERTION_SPACING would insert sps/pps ...

**Review by pkviet — CHANGES_REQUESTED** [review]
please make the changes I suggested.
Also in order to help us to reproduce the issue, please provide us with info regarding your GPU (which model).

**Comment by shinji3** [comment]
I'm using a Radeon RX550

**Comment by pkviet** [comment]
a single commit is enough, no need to make 3 commits

**Comment by flaeri** [comment]
Tested the PR (37c8da56d) on texture and fallback, h264 and HEVC. AMD r5 4500U APU, and I am able to playback the local ts files as standalone.

I was not able to prior (on release).

**Comment by pkviet** [comment]
LGTM; resolve the conversations and we should be good

File: plugins/obs-ffmpeg/texture-amf.cpp
Comment (RytoEX) [reply plugins/obs-ffmpeg/texture-amf.cpp]: Resolved per off-thread request.

**Comment by RytoEX** [comment]
Waiting on blocking conversations to be resolved before merging.

**Comment by RytoEX** [comment]
> also, for future reference, @RytoEX unearthed this discussion :[GPUOpen-LibrariesAndSDKs/AMF#293](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/issues/293) Without these infos, it would have been impossible to guess that HEADER_INSERTION_SPACING would insert sps/pps ...

Follow-up that the [v1.4.24 AMF_Video_Encode_API.pdf](https://raw.githubusercontent.com/GPUOpen-LibrariesAndSDKs/AMF/v1.4.24/amf/doc/AMF_Video_Encode_API.pdf) mentions:
> ** HeaderInsertionSpacing: Every IDR frame has SPS and PPS regardless of default value of HeaderInsertionSpacing
per VCE logic.

But [AMD says](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/issues/293#issuecomment-1051301399):
> if you want that SPS/PPS inserted with every IDR controlled by AMF_VIDEO_ENCODER_IDR_PERIOD please set AMF_VIDEO_ENCODER_HEADER_INSERTION_SPACING to the same value.

Those seem to say the opposite thing to me, but perhaps I've misunderstood.

Anyway, I _think_ this is okay to land. Will look over it again. Other approvals also welcome.

**Review by RytoEX — APPROVED** [review]
Looks fine at a glance.

**Comment by jp9000** [comment]
Do the other encoders have this behavior by default?

**Comment by pkviet** [comment]
> Do the other encoders have this behavior by default?

i checked that it is not the case with nvenc h264/hevc & x264. The ts recording works fine. It's only with amf oddly enough.

**Comment by jp9000** [comment]
> i checked that it is not the case with nvenc h264/hevc & x264. The ts recording works fine. It's only with amf oddly enough.

Respectfully, "oddly enough" is an uncomfortable answer. It would be nice to know why specifically this only affects AMF and not nvenc/x264.

**Comment by flaeri** [comment]
I cannot test nvenc atm, due to not having it accessible for the next few wekks.

I can however confirm that x264 TS segments (from m3u8) does contain PPS and SPS by default.

If someone would be so kind as to drop me said ts segment from nvenc, I'll be happy to check.

AMF as establish does not contain them in the current release build.
image

**Comment by jp9000** [comment]
I wonder if I should just turn on header repetition by default for amf.

**Comment by derrod** [comment]
> I wonder if I should just turn on header repetition by default for amf.

I believe that would result in issues with other container formats such as MP4 or MKV.

**Comment by shinji3** [comment]
>I wonder if I should just turn on header repetition by default for amf.

flv, mp4, mov, mkv, ts, m3u8 could be recorded without issues.

**Pull request merged**
